File: ios/StarXpandCommand/PreSettingBuilderWrapper.m
#import "PreSettingBuilderWrapper.h"
#import "StarObjectManager.h"
#import "StarXpandCommandBuilderWrapper.h"

@import StarIO10;

@interface PreSettingBuilderWrapper()

@property(assign, nonatomic) StarObjectManager *objManager;

@end


@implementation PreSettingBuilderWrapper

- (instancetype)init
{
    self = [super init];
    if (self) {
        _objManager = StarObjectManager.sharedManager;
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

RCT_EXPORT_MODULE()

RCT_REMAP_METHOD(init,
                 createWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandPreSettingBuilder *builder = [[STARIO10StarXpandCommandPreSettingBuilder alloc] init];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to create object.", nil);
        return;
    }

    NSString *objID = [_objManager add:builder];
    
    resolve(objID);
}

RCT_REMAP_METHOD(dispose,
                 disposeWithObjectIdentifier:(nonnull NSString *)objID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    [_objManager remove:objID];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addPresenterSetting,
                 addPresenterSettingWithObjectIdentifier:(nonnull NSString *)objID
                 presenterSettingBuilderIdentifier:(nonnull NSString *)presenterSettingBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandPreSettingBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandPreSettingBuilder *presenterSettingBuilder = [_objManager getObject:presenterSettingBuilderID];
    
    if (presenterSettingBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addPresenterSetting:presenterSettingBuilder];

    resolve(nil);
}

RCT_REMAP_METHOD(addBezelSetting,
                 addBezelSettingWithObjectIdentifier:(nonnull NSString *)objID
                 bezelSettingBuilderIdentifier:(nonnull NSString *)bezelSettingBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandPreSettingBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandBezelSettingBuilder *bezelSettingBuilder = [_objManager getObject:bezelSettingBuilderID];
    if (bezelSettingBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addBezelSetting:bezelSettingBuilder];

    resolve(nil);
}

@end

File: android/src/main/java/com/stario10module/PreSettingBuilderWrapper.kt
package com.stario10module

import com.facebook.react.bridge.*
import com.starmicronics.stario10.starxpandcommand.BezelSettingBuilder
import com.starmicronics.stario10.starxpandcommand.PreSettingBuilder
import com.starmicronics.stario10.starxpandcommand.PresenterSettingBuilder

class PreSettingBuilderWrapper internal constructor(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "PreSettingBuilderWrapper"
    }

    @ReactMethod
    fun init(promise: Promise) {
        val builder = PreSettingBuilder()
        val identifier = InstanceManager.set(builder)

        promise.resolve(identifier)
    }

    @ReactMethod
    fun dispose(identifier: String, promise: Promise) {
        InstanceManager.remove(identifier)
        promise.resolve(0)
    }

    @ReactMethod
    fun addPresenterSetting(identifier: String, presenterSettingBuilderIdentifier: String, promise: Promise) {
        val builder = InstanceManager.get(identifier)
        val presenterSettingBuilder = InstanceManager.get(presenterSettingBuilderIdentifier)

        if (builder is PreSettingBuilder && presenterSettingBuilder is PresenterSettingBuilder) {
            builder.addPresenterSetting(presenterSettingBuilder)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addBezelSetting(identifier: String, bezelSettingBuilderIdentifier: String, promise: Promise) {
        val builder = InstanceManager.get(identifier)
        val bezelSettingBuilder = InstanceManager.get(bezelSettingBuilderIdentifier)

        if (builder is PreSettingBuilder && bezelSettingBuilder is BezelSettingBuilder) {
            builder.addBezelSetting(bezelSettingBuilder)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }
}